File: Leetcode/Java/24. Swap Nodes in Pairs/solution.java
// https://leetcode.com/problems/swap-nodes-in-pairs/description/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode prev = new ListNode(-1, head);
        ListNode root = prev;

        while (head != null && head.next != null){
            ListNode nextNode = head.next;
            prev.next = nextNode;
            head.next = nextNode.next;
            nextNode.next = head;

            prev = head;
            head = head.next;
        }

        return root.next;
    }
}

File: Leetcode/Kotlin/24. Swap Nodes in Pairs/solution.kt
// https://leetcode.com/problems/swap-nodes-in-pairs/description/

/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun swapPairs(head: ListNode?): ListNode? {
        val root = ListNode(-1)
        root.next = head
        var prev = root
        var cur = head

        while (cur?.next != null){
            var nextNode = cur.next
            cur.next = nextNode.next
            nextNode.next = cur
            prev.next = nextNode

            prev = cur
            cur = cur.next
        }

        return root.next
    }
}